core/example/longrange/spme.cpp

```diff
@@ -358,15 +359,27 @@ double TPME::compute( ParticleList& particles, ParticleList& mesh, double lx, do
                   double m2 = (mx*mx + my*my + mz*mz);//Unnecessary extra variable
 
                   //Calculate BC. Why store the imag part at all?
+                  #ifdef Cabana_ENABLE_Cuda
                   BC[idx].x =  TPME::oneDeuler(kx,meshwidth) * TPME::oneDeuler(ky,meshwidth)
                                * TPME::oneDeuler(kz,meshwidth)
                                * exp( -PI*PI*m2 / (alpha*alpha) ) / (PI * lx*ly*lz * m2 );
                   BC[idx].y = 0.0;//imag part
+                  #else
+                  BC[idx][0] =  TPME::oneDeuler(kx,meshwidth) * TPME::oneDeuler(ky,meshwidth)
+                               * TPME::oneDeuler(kz,meshwidth)
+                               * exp( -PI*PI*m2 / (alpha*alpha) ) / (PI * lx*ly*lz * m2 );
+                  BC[idx][1] = 0.0;//imag part
+                  #endif
               }
               else
               {
+                  #ifdef Cabana_ENABLE_Cuda
                   BC[idx].x = 0.0;
                   BC[idx].y = 0.0;//set origin element to zero
+                  #else
+                  BC[idx][0] = 0.0;
+                  BC[idx][1] = 0.0;//set origin element to zero
+                  #endif
```

Comment:
Did you guys consider encapsulating the element access to avoid the proliferation of preprocessor directives?

Reply:
Yes, our to-do list includes wrapping all FFT stuff to mostly remove those directives cluttering the code.

Reply:
```c++
#ifdef Cabana_ENABLE_Cuda
inline double& real(cufftDoubleComplex& c) { return c.x; }
inline double const& real(cufftDoubleComplex const& c) { return c.x; }
inline double& imag(cufftDoubleComplex& c) { return c.y; }
inline double const& imag(cufftDoubleComplex const& c) { return c.y; }
#else
inline double& real(fftw_complex& c) { return c[0]; }
inline double const& real(fftw_complex const& c) { return c[0] }
inline double& imag(fftw_complex& c) { return c[1]; }
inline double const& imag(fftw_complex const& c) { return c[1]; }
#endif
```

Reply:
Anyway this belong to another PR.